Change p2pserver connection cache from slice to set

rongyi opened the pull request:
1. the time complexity of searching in slice is O(n) so change from slice to map give a better performace.
2. ConnectingNodes field in NetServer need not to export, all access to this field is from the 4 API(AddOutConnectingList/RemoveFromConnectingList/GetOutConnectingListLen/ IsAddrFromConnecting) so lowercase this field.
3. the other two cache(inConnRecord/outConnRecord) are similar with ConnectingNodes
4. Add testcase to cover those change.

rongyi committed:
Change cache from slice to set

rongyi committed:
Update p2pserver testcase

laizy commented on p2pserver/common/set/string_set.go:
```diff
@@ -0,0 +1,189 @@
+	"sort"
+)
+
+type Empty struct{}
```
internal type should not be public.

laizy commented on p2pserver/common/set/string_set.go:
```diff
@@ -0,0 +1,189 @@
+type Empty struct{}
+
+// sets.String is a set of strings, implemented via map[string]struct{} for minimal memory consumption.
+type String map[string]Empty
```
It's an uncomfortable name. I prefer `Strings` or `StringSet`

rongyi replied:
Thanks for pointing out

laizy commented on p2pserver/common/set/string_set.go:
```diff
@@ -0,0 +1,189 @@
+
+// StringKeySet creates a String from a keys of a map[string](? extends interface{}).
+// If the value passed in is not actually a map, this will panic.
+func StringKeySet(theMap interface{}) String {
```
remove all unused functions

rongyi replied:
Done

laizy commented on p2pserver/common/set/string_set.go:
```diff
@@ -0,0 +1,189 @@
+// s2 = {a1, a2, a4, a5}
+// s1.Difference(s2) = {a3}
+// s2.Difference(s1) = {a4, a5}
+func (s String) Difference(s2 String) String {
```
I notice slightly inconsistent semantics: some function like `Delete` and `Insert` modify the original set and return it. but this function and the following one keep the original set unchanged and return a new one. I suggest removing the return value from `Delete` and `Insert` to avoid misuse. 
add unit test or remove this function since it is not used.

rongyi replied:
Remove unused API

rongyi committed:
Add license header

rongyi committed:
Drop unused string set api and review change:

1. String ==> StringSet
2. Add licence header for new file
3. Drop unused set api
